amazons3/amazons3_utils.bal

```diff
@@ -21,158 +20,332 @@ import ballerina/encoding;
+
+    // Generate string to sign.
+    string stringToSign = generateStringToSign(amzDateStr, shortDateStr,region, canonicalRequest);
+    
+    // Construct authorization signature string.
+    string authHeader = constructAuthSignature(accessKeyId, secretAccessKey, shortDateStr, region, signedHeaders, 
+                            stringToSign);
+    // Set authorization header.
+    request.setHeader(AUTHORIZATION, authHeader);
+}
+
+function setResponseError(string errorMessage) returns error {
+    error err = error(AMAZONS3_ERROR_CODE, { message : errorMessage });
+    return err;
+}
+
+# Funtion to generate the date strings.
+# 
+# + return - amzDate string and short date string.
+function generateDateString() returns (string, string) {
     time:Time|error time = time:toTimeZone(time:currentTime(), "GMT");
+    string amzDateStr;
+    string shortDateStr;
     if (time is time:Time) {
         string|error amzDate = time:format(time, ISO8601_BASIC_DATE_FORMAT);
         string|error shortDate = time:format(time, SHORT_DATE_FORMAT);
         if (amzDate is string) {
             amzDateStr = amzDate;
+
         } else {
-            return amzDate;
+            panic amzDate;
         }
         if (shortDate is string) {
             shortDateStr = shortDate;
         } else {
-            return shortDate;
+            panic shortDate;
         }
     } else {
-        return time;
+        panic time;
     }
+    return (amzDateStr, shortDateStr);
+}
+
+# Function to generate string to sign.
+# 
+# + amzDateStr - amzDate string.
+# + shortDateStr - Short date string.
+# + region - Endpoint region.
+# + canonicalRequest - Generated canonical request.
+# 
+# + return - String to sign.
+function generateStringToSign(string amzDateStr, string shortDateStr, string region, string canonicalRequest) 
+                            returns string{
+    //Start creating the string to sign
+    string stringToSign = string `${AWS4_HMAC_SHA256}\n${amzDateStr}\n${shortDateStr}`;
+    stringToSign = string `${stringToSign}/${region}/${SERVICE_NAME}/${TERMINATION_STRING}\n`;
+    stringToSign = stringToSign + encoding:encodeHex(crypto:hashSha256(canonicalRequest.toByteArray(UTF_8))).toLower();
+    return stringToSign;
+
+}
 
-    request.setHeader(X_AMZ_DATE, amzDateStr);
-    canonicalRequest = httpVerb;
-    canonicalRequest = canonicalRequest + "\n";
+# Function to get canonical URI.
+#
+# + requestURI - Request URI. 
+# 
+# + return - Return encoded request URI.
+function getCanonicalURI(string requestURI) returns string {
+    string encodedrequestURIValue;
     var value = http:encode(requestURI, UTF_8);
     if (value is string) {
-        encodedrequestURIValue = value;
+        encodedrequestURIValue = value.replace("%2F", "/");
     } else {
         error err = error(AMAZONS3_ERROR_CODE, { message: "Error occurred when converting to string"});
         panic err;
```

Comment:
```suggestion
        panic error(AMAZONS3_ERROR_CODE, { message: "Error occurred when converting to string"});;
```
would also work.

Shouldn't we propagate the original error (or at least some information from it), maybe as part of the error detail? If we don't, the user may not be able to figure out what went wrong right?

Reply:
If we are not using the error from encode, and we return at L130, we can remove the else block and move this logic out?